### DISCUSSION TITLE
Implementing a second optimizer for a specific variable deeper in the code.

### DISCUSSION THREAD

**Discussion by a-froghyar** [coqui-ai/TTS#406] [General Q&A]
Hey,

In the context of my thesis, I'm implementing [Capacitron](https://arxiv.org/abs/1906.03402), a variational autoencoder extension of the GST-Taco1 architecture already present in coqui TTS.

In this paper, the authors define a new loss function, eq.9 in the paper above. This new loss function is then passed onto two separate optimizers, one is the basic ADAM optimizer for the model parameters, and one is a separate SGD optimizer that is only accounting the right term in eq.9 above. It's basically a lagrange-like double optimisation routine, whereby the 'lagrange multiplier' is optimized by a separate optimizer.

My questions to the community are:

1. How does the current optimizer work?

The existing ADAM optimizer (`RAdam()` in the code) receives as input the model parameters fetched from the `setup_model()` method above. The `setup_model()` method returns the big Tactron model class - how is this optimizer applied to the loss functions, as in how does it know that it needs to do backprop based on the loss function we defined?

2. How I should go on about passing this specific variable deep in the code into the optimizer?

I'm defining the second optimizer in `train_tacotron.py` in the `main()` function around line 550. I'd like to pass only one specific variable I've initialised in one of my modules deep in the code to this optimizer. Is there a way that I can fetch this variable somehow from the big model that's returned by `setup_model()`?

Many thanks for the help!

**Comment by erogol** [comment]
It is more of a Pytorch question than it is for TTS but I try to give my two cents.

What you can do is that you can create another optimizer class that uses two optimizers together optimizing respective parameters. That way you don't need to change the TTS training code too much except changing the optimizer in use to your own implementation. 

the custom Radam optimizer in TTS is a good example to show how you can implement your own. 
https://github.com/coqui-ai/TTS/blob/main/TTS/utils/radam.py
